### ISSUE TITLE
Checksums differ with starterkit - even though doctypes are identical

### ISSUE THREAD

**Issue by hjaltedaniel-umbraco** [umbraco/Umbraco.Cloud.Issues#290]
When using the Starter Kit on a project on Cloud, and trying to restore you currently get a "Schema mismatch" even though the doctypes, schema etc. is identical.
Reproduction
------------

If you're filing a bug, please describe how to reproduce it. Include as much
relevant information as possible, such as:

### Bug summary
When restoring content with the "Full restore" functionality you get a schema mismatch.
It seems that the order of the allowed child doctypes is the one causing issues. Moving the order, be that in uda file or in backoffice, does not fix it.

### Specifics
Version CMS: 8.9.1
Version Deploy: 3.5.3

This is the error:
```
Document type Home with alias 'home'
umb://document-type/eecddf81abcf47ae84568fab70c45b63
No differences found, but checksums differ.
Source checksum json: {"DefaultTemplate":"umb://template/800bcd8733844affa96e67c050323eb2","AllowedTemplates":["umb://template/800bcd8733844affa96e67c050323eb2"],"Icon":"icon-home color-black","Thumbnail":"folder.png","Description":null,"IsContainer":false,"Permissions":{"AllowVaryingByCulture":false,"AllowVaryingBySegment":false,"AllowedAtRoot":true,"IsElementType":false,"AllowedChildContentTypes":["umb://document-type/3e16fc3f16df4abda8d25f2123d1c235","umb://document-type/7c2ea25fb5f54afb8972db6a7433a3e8","umb://document-type/3a6b71355d0f438a84ff8090a907cbdb","umb://document-type/424f4da4ba3a434e9688350dee9db1e2","umb://document-type/4957a80c26dd4bc8a47fb3768a1f4b51"]},"Parent":null,"CompositionContentTypes":[],"PropertyGroups":[{"Key":"2f72d77e-e5a5-43c8-b6d3-750f1e4326ab","Name":"Design","SortOrder":3,"PropertyTypes":[{"Key":"15adfc52-84ac-4a90-bb38-bc1819958bea","Alias":"sitename","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":"Used on the homepage as well as the title and social cards","Mandatory":true,"Name":"Sitename","SortOrder":3,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"47e3deed-f6b6-436c-8ab4-2a60f4a259fa","Alias":"SiteLogo","DataType":"umb://data-type/135d60e064d949edab08893c9ba44ae5","Description":"Optional. If you add a logo it'll be used in the upper left corner instead of the site name. Make sure to use a transparent logo for best results","Mandatory":false,"Name":"Logo","SortOrder":4,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"520c09c2-77f7-4c0b-be31-2647f5f67c68","Alias":"colorTheme","DataType":"umb://data-type/71548df5836c41f896ac5b98fe5e2e19","Description":"","Mandatory":true,"Name":"Color Theme","SortOrder":2,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"77c2ba60-6e9f-4130-9fed-bd27c211c4c8","Alias":"font","DataType":"umb://data-type/fff3c3603f9045b3a554e29ca72cdce4","Description":"","Mandatory":true,"Name":"Font","SortOrder":1,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"9e64eba8-e6ff-43b2-94a6-18cbf35bb63a","Alias":"HeroBackgroundImage","DataType":"umb://data-type/135d60e064d949edab08893c9ba44ae5","Description":"Spice up the homepage by adding a beautiful photo that relates to your business","Mandatory":true,"Name":"Hero Background","SortOrder":0,"Validation":null,"VariesByCulture":false,"VariesBySegment":false}]},{"Key":"509ca594-bdb9-41ed-8994-b8a5395aed0b","Name":"Content","SortOrder":1,"PropertyTypes":[{"Key":"c64aa049-a0e7-42c2-9b61-453d8831f53c","Alias":"bodyText","DataType":"umb://data-type/34a7f8f5a84e439b9322466ff7ed8866","Description":null,"Mandatory":false,"Name":"Content","SortOrder":0,"Validation":null,"VariesByCulture":false,"VariesBySegment":false}]},{"Key":"941d9319-5792-4c12-afa1-ad960402e580","Name":"Footer","SortOrder":2,"PropertyTypes":[{"Key":"69cafe97-b3f4-4e5b-81e5-f943556d10bd","Alias":"footerCTACaption","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":"Caption on the Call To Action Button","Mandatory":false,"Name":"Call To Action Caption","SortOrder":2,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"6cec5cc5-593a-4fe9-8a1c-e9972b5e8ca2","Alias":"footerAddress","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":null,"Mandatory":true,"Name":"Address","SortOrder":4,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"81c712d9-5f97-4713-939a-cb5a7aade55d","Alias":"footerDescription","DataType":"umb://data-type/c6bac0dd4ab945b18e30e4b619ee5da3","Description":null,"Mandatory":false,"Name":"Description","SortOrder":1,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"9fcb4d5d-ce73-483a-a41c-260ecc3c1c33","Alias":"footerHeader","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":null,"Mandatory":false,"Name":"Header","SortOrder":0,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"e053890b-b085-4cc5-ad79-b20de8ba5006","Alias":"FooterCtalink","DataType":"umb://data-type/3ece86aa61ad45d5b1dd8f02f25df949","Description":null,"Mandatory":true,"Name":"Call To Action Link","SortOrder":3,"Validation":null,"VariesByCulture":false,"VariesBySegment":false}]},{"Key":"a0e659c3-14ef-46e2-8423-7399de369331","Name":"Hero","SortOrder":0,"PropertyTypes":[{"Key":"6dec1d58-8724-430b-974f-6c070263bce2","Alias":"heroDescription","DataType":"umb://data-type/c6bac0dd4ab945b18e30e4b619ee5da3","Description":null,"Mandatory":true,"Name":"Description","SortOrder":1,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"d75571db-6450-4faa-b456-dfff44a8c24c","Alias":"heroCTACaption","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":"The caption on the button","Mandatory":true,"Name":"Call To Action Caption","SortOrder":2,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"dec866ab-0694-4ed0-afcc-2e5c6604b41e","Alias":"HeroCtalink","DataType":"umb://data-type/3ece86aa61ad45d5b1dd8f02f25df949","Description":null,"Mandatory":true,"Name":"Call To Action Link","SortOrder":3,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"ecee048f-5b28-4ee1-ad9b-1303ceb74bff","Alias":"heroHeader","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":"This is the main headline for the hero area on the Homepage","Mandatory":true,"Name":"Header","SortOrder":0,"Validation":null,"VariesByCulture":false,"VariesBySegment":false}]}],"PropertyTypes":[],"Udi":"umb://document-type/eecddf81abcf47ae84568fab70c45b63","Dependencies":[{"Udi":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Ordering":true,"Mode":0},{"Udi":"umb://data-type/135d60e064d949edab08893c9ba44ae5","Ordering":true,"Mode":0},{"Udi":"umb://data-type/34a7f8f5a84e439b9322466ff7ed8866","Ordering":true,"Mode":0},{"Udi":"umb://data-type/3ece86aa61ad45d5b1dd8f02f25df949","Ordering":true,"Mode":0},{"Udi":"umb://data-type/71548df5836c41f896ac5b98fe5e2e19","Ordering":true,"Mode":0},{"Udi":"umb://data-type/c6bac0dd4ab945b18e30e4b619ee5da3","Ordering":true,"Mode":0},{"Udi":"umb://data-type/fff3c3603f9045b3a554e29ca72cdce4","Ordering":true,"Mode":0},{"Udi":"umb://document-type/3a6b71355d0f438a84ff8090a907cbdb","Ordering":false,"Mode":0},{"Udi":"umb://document-type/3e16fc3f16df4abda8d25f2123d1c235","Ordering":false,"Mode":0},{"Udi":"umb://document-type/424f4da4ba3a434e9688350dee9db1e2","Ordering":false,"Mode":0},{"Udi":"umb://document-type/4957a80c26dd4bc8a47fb3768a1f4b51","Ordering":false,"Mode":0},{"Udi":"umb://document-type/7c2ea25fb5f54afb8972db6a7433a3e8","Ordering":false,"Mode":0},{"Udi":"umb://template/800bcd8733844affa96e67c050323eb2","Ordering":true,"Mode":0}],"Name":"Home","Alias":"home","__type":"Umbraco.Deploy,Umbraco.Deploy.Artifacts.ContentType.DocumentTypeArtifact"}
Target checksum json: {"DefaultTemplate":"umb://template/800bcd8733844affa96e67c050323eb2","AllowedTemplates":["umb://template/800bcd8733844affa96e67c050323eb2"],"Icon":"icon-home color-black","Thumbnail":"folder.png","Description":null,"IsContainer":false,"Permissions":{"AllowVaryingByCulture":false,"AllowVaryingBySegment":false,"AllowedAtRoot":true,"IsElementType":false,"AllowedChildContentTypes":["umb://document-type/3a6b71355d0f438a84ff8090a907cbdb","umb://document-type/3e16fc3f16df4abda8d25f2123d1c235","umb://document-type/424f4da4ba3a434e9688350dee9db1e2","umb://document-type/4957a80c26dd4bc8a47fb3768a1f4b51","umb://document-type/7c2ea25fb5f54afb8972db6a7433a3e8"]},"Parent":null,"CompositionContentTypes":[],"PropertyGroups":[{"Key":"2f72d77e-e5a5-43c8-b6d3-750f1e4326ab","Name":"Design","SortOrder":3,"PropertyTypes":[{"Key":"15adfc52-84ac-4a90-bb38-bc1819958bea","Alias":"sitename","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":"Used on the homepage as well as the title and social cards","Mandatory":true,"Name":"Sitename","SortOrder":3,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"47e3deed-f6b6-436c-8ab4-2a60f4a259fa","Alias":"SiteLogo","DataType":"umb://data-type/135d60e064d949edab08893c9ba44ae5","Description":"Optional. If you add a logo it'll be used in the upper left corner instead of the site name. Make sure to use a transparent logo for best results","Mandatory":false,"Name":"Logo","SortOrder":4,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"520c09c2-77f7-4c0b-be31-2647f5f67c68","Alias":"colorTheme","DataType":"umb://data-type/71548df5836c41f896ac5b98fe5e2e19","Description":"","Mandatory":true,"Name":"Color Theme","SortOrder":2,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"77c2ba60-6e9f-4130-9fed-bd27c211c4c8","Alias":"font","DataType":"umb://data-type/fff3c3603f9045b3a554e29ca72cdce4","Description":"","Mandatory":true,"Name":"Font","SortOrder":1,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"9e64eba8-e6ff-43b2-94a6-18cbf35bb63a","Alias":"HeroBackgroundImage","DataType":"umb://data-type/135d60e064d949edab08893c9ba44ae5","Description":"Spice up the homepage by adding a beautiful photo that relates to your business","Mandatory":true,"Name":"Hero Background","SortOrder":0,"Validation":null,"VariesByCulture":false,"VariesBySegment":false}]},{"Key":"509ca594-bdb9-41ed-8994-b8a5395aed0b","Name":"Content","SortOrder":1,"PropertyTypes":[{"Key":"c64aa049-a0e7-42c2-9b61-453d8831f53c","Alias":"bodyText","DataType":"umb://data-type/34a7f8f5a84e439b9322466ff7ed8866","Description":null,"Mandatory":false,"Name":"Content","SortOrder":0,"Validation":null,"VariesByCulture":false,"VariesBySegment":false}]},{"Key":"941d9319-5792-4c12-afa1-ad960402e580","Name":"Footer","SortOrder":2,"PropertyTypes":[{"Key":"69cafe97-b3f4-4e5b-81e5-f943556d10bd","Alias":"footerCTACaption","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":"Caption on the Call To Action Button","Mandatory":false,"Name":"Call To Action Caption","SortOrder":2,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"6cec5cc5-593a-4fe9-8a1c-e9972b5e8ca2","Alias":"footerAddress","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":null,"Mandatory":true,"Name":"Address","SortOrder":4,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"81c712d9-5f97-4713-939a-cb5a7aade55d","Alias":"footerDescription","DataType":"umb://data-type/c6bac0dd4ab945b18e30e4b619ee5da3","Description":null,"Mandatory":false,"Name":"Description","SortOrder":1,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"9fcb4d5d-ce73-483a-a41c-260ecc3c1c33","Alias":"footerHeader","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":null,"Mandatory":false,"Name":"Header","SortOrder":0,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"e053890b-b085-4cc5-ad79-b20de8ba5006","Alias":"FooterCtalink","DataType":"umb://data-type/3ece86aa61ad45d5b1dd8f02f25df949","Description":null,"Mandatory":true,"Name":"Call To Action Link","SortOrder":3,"Validation":null,"VariesByCulture":false,"VariesBySegment":false}]},{"Key":"a0e659c3-14ef-46e2-8423-7399de369331","Name":"Hero","SortOrder":0,"PropertyTypes":[{"Key":"6dec1d58-8724-430b-974f-6c070263bce2","Alias":"heroDescription","DataType":"umb://data-type/c6bac0dd4ab945b18e30e4b619ee5da3","Description":null,"Mandatory":true,"Name":"Description","SortOrder":1,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"d75571db-6450-4faa-b456-dfff44a8c24c","Alias":"heroCTACaption","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":"The caption on the button","Mandatory":true,"Name":"Call To Action Caption","SortOrder":2,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"dec866ab-0694-4ed0-afcc-2e5c6604b41e","Alias":"HeroCtalink","DataType":"umb://data-type/3ece86aa61ad45d5b1dd8f02f25df949","Description":null,"Mandatory":true,"Name":"Call To Action Link","SortOrder":3,"Validation":null,"VariesByCulture":false,"VariesBySegment":false},{"Key":"ecee048f-5b28-4ee1-ad9b-1303ceb74bff","Alias":"heroHeader","DataType":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Description":"This is the main headline for the hero area on the Homepage","Mandatory":true,"Name":"Header","SortOrder":0,"Validation":null,"VariesByCulture":false,"VariesBySegment":false}]}],"PropertyTypes":[],"Udi":"umb://document-type/eecddf81abcf47ae84568fab70c45b63","Dependencies":[{"Udi":"umb://data-type/0cc0eba1996042c9bf9b60e150b429ae","Ordering":true,"Mode":0},{"Udi":"umb://data-type/135d60e064d949edab08893c9ba44ae5","Ordering":true,"Mode":0},{"Udi":"umb://data-type/34a7f8f5a84e439b9322466ff7ed8866","Ordering":true,"Mode":0},{"Udi":"umb://data-type/3ece86aa61ad45d5b1dd8f02f25df949","Ordering":true,"Mode":0},{"Udi":"umb://data-type/71548df5836c41f896ac5b98fe5e2e19","Ordering":true,"Mode":0},{"Udi":"umb://data-type/c6bac0dd4ab945b18e30e4b619ee5da3","Ordering":true,"Mode":0},{"Udi":"umb://data-type/fff3c3603f9045b3a554e29ca72cdce4","Ordering":true,"Mode":0},{"Udi":"umb://document-type/3a6b71355d0f438a84ff8090a907cbdb","Ordering":false,"Mode":0},{"Udi":"umb://document-type/3e16fc3f16df4abda8d25f2123d1c235","Ordering":false,"Mode":0},{"Udi":"umb://document-type/424f4da4ba3a434e9688350dee9db1e2","Ordering":false,"Mode":0},{"Udi":"umb://document-type/4957a80c26dd4bc8a47fb3768a1f4b51","Ordering":false,"Mode":0},{"Udi":"umb://document-type/7c2ea25fb5f54afb8972db6a7433a3e8","Ordering":false,"Mode":0},{"Udi":"umb://template/800bcd8733844affa96e67c050323eb2","Ordering":true,"Mode":0}],"Name":"Home","Alias":"home","__type":"Umbraco.Deploy,Umbraco.Deploy.Artifacts.ContentType.DocumentTypeArtifact"}
```

### Steps to reproduce
1. Install Starter Kit
2. Take a local clone
3. Click "Full restore" - you will get the schema mismatch

### Expected result
A full restore, without issues.

### Actual result
Schema mismatch (workaround is to use partial restore)

**Comment by ronaldbarendse** [comment]
If the actual UDA-files on the local and Live environment are exactly the same, it seems the sort order of the allow child node types isn't applied when comparing checksums and running a deploy-export, see https://github.com/umbraco/Umbraco.Cloud.Issues/issues/203#issuecomment-740638900.

Another workaround is to (temporarly) disable document types from deploy (see linked issue).

**Comment by rustybox** [comment]
+ Follow https://our.umbraco.com/documentation/Umbraco-Cloud/Getting-started/
+ Create brand new trial instance to evaluate product, use starter kit
+ Step 1 after initial deployment is setup for local development
+ Click restore button and experience giant error page

Workarounds aren't appropriate here, something is very wrong for the new customer story

**Comment by Jay-umbr** [comment]
I have seen this happen on a project not using the starter kit. I do not think that is limited just to our package 🤔

**Comment by orsinic** [comment]
We've see this on a custom site (not a starter kit) this week as well. Content restore to local failed with the same error as above "No differences found, but checksums differ." for some doc types.

The uda's for source (local) and target (Umbraco Cloud Development environment) were exactly the same, but comparing the source and target json listed in the deploy error message showed differences between the order of the allowed child doctypes. It's unclear from where the json is pulled since they do not correspond 100% to the uda files.
We're using the latest version of Umbraco (8.9.1) and Deploy (3.5.3).

What does work in our case is starting from scratch locally (by removing the database file Umbraco.mdf) and doing a full restore from Development. So it seems there's a mismatch in the config of items (doc types in this case) in the local database and the uda files.

**Comment by RyuLindow** [comment]
Just had it on a project that was made from a baseline and it threw the same error as in the thread.

Partial restore "solved it".

It's using deploy 3.5.3

**Comment by FransdeJong** [comment]
I ran into this issue today. I didn't use the starter kit but the project was created from a baseline.
Full restore goes wrong but partial restore goes alright.

I'm running Umbraco 8.9.1 and deploy 3.5.3

**Comment by HamDerAndrew** [comment]
Ran into this issue today. The project I worked on did not include a starter kit.
Dev, staging and live environments were in sync.

As mentioned, using Partial Restore as a workaround worked for me, but it's not intuitive.

**Comment by ComplX1** [comment]
We encountered this same issue and found that oftentimes after building in Visual Studio we would have template files which looked identical but git was identifying changes. We found the character encoding was different between the files. so even thought they looked the same the checksum was different. we ended up going through the effected files. comparing the encoding from the files generated from umbraco cloud. and from then on we didnt have any issues with those files.

**Comment by rustybox** [comment]
Thing the mysterious hash differences just got explained :)

**Comment by MichaelNielsenDK** [comment]
Also encountered this, I am not using a starter kit. 

The issue for me, was that in the checksum json, the sorting of the allowed child nodes was different between source and target, even though they were the same in the backend on local and live environements. 

Have been in contact with Umbraco Support, which says the issue will be fixed in next version of Umbraco.Deploy. 

They gave me a temporary fix, which is attached here. Just drop it in the bin folder on both environments, and run `echo > deploy-clearsignatures`

[Umbraco.Deploy.zip](https://github.com/umbraco/Umbraco.Cloud.Issues/files/5951615/Umbraco.Deploy.zip)

**Comment by ronaldbarendse** [comment]
@ComplX1 Although different encodings could indeed result in mismatching checksums, in this case that's not the reason: it clearly uses a different order in `AllowedChildContentTypes`.

@MichaelNielsenDK Great to see you've received a temporary fix from Umbraco support, but I would be hesitant to apply unreleased hotfixes from an unofficial source and rather temporarily disable deploying document types (see https://github.com/umbraco/Umbraco.Cloud.Issues/issues/203#issuecomment-740638900) while waiting for the next patch release 😇

**Comment by MichaelNielsenDK** [comment]
@ronaldbarendse It is a sound approach to not accept candy from strangers 😅 

I figured I would just make it available, then it's up to each individual, if they want to use it or not 🤷‍♂️

**Comment by nul800sebastiaan** [comment]
This is fixed in Deploy 3.5.4, rolling out in a few days.
